contracts/multisig-prover/src/execute.rs

```diff
@@ -126,31 +128,123 @@ pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractEr
         pub_keys.push(pub_key);
     }
 
-    let new_worker_set = WorkerSet::new(snapshot, pub_keys, env.block.height)?;
+    Ok(WorkersInfo {
+        snapshot,
+        pubkeys_by_participant: participants.into_iter().zip(pub_keys).collect(),
+    })
+}
+
+pub fn update_worker_set(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    let workers_info = get_workers_info(&deps, &env, &config)?;
+
+    let new_worker_set = WorkerSet::new(
+        workers_info.pubkeys_by_participant,
+        workers_info.snapshot.quorum.into(),
+        env.block.height,
+    )?;
 
-    let cur_worker_set = CURRENT_WORKER_SET.load(deps.storage)?;
+    let cur_worker_set = CURRENT_WORKER_SET.may_load(deps.storage)?;
 
-    if !should_update_worker_set(
-        &new_worker_set,
-        &cur_worker_set,
-        config.worker_set_diff_threshold as usize,
-    ) {
-        return Err(ContractError::WorkerSetUnchanged);
+    let key_id = new_worker_set.id();
+
+    match cur_worker_set {
+        None => {
+            // if no worker set, just store it and return
+            CURRENT_WORKER_SET.save(deps.storage, &new_worker_set)?;
+
+            KEY_ID.save(deps.storage, &key_id)?;
+
+            let key_gen_msg = multisig::msg::ExecuteMsg::KeyGen {
+                key_id,
+                snapshot: workers_info.snapshot,
+                pub_keys_by_address: new_worker_set
+                    .signers
+                    .into_iter()
+                    .map(|signer| {
+                        (
+                            signer.address.to_string(),
+                            (KeyType::Ecdsa, signer.pub_key.as_ref().into()),
+                        )
+                    })
+                    .collect(),
+            };
+
+            Ok(Response::new().add_message(wasm_execute(config.multisig, &key_gen_msg, vec![])?))
+        }
+        Some(cur_worker_set) => {
+            if !should_update_worker_set(
+                &new_worker_set,
+                &cur_worker_set,
+                config.worker_set_diff_threshold as usize,
+            ) {
+                return Err(ContractError::WorkerSetUnchanged);
+            }
+
+            NEXT_WORKER_SET.save(
```

Comment:
good one, I believe we need to revert because if I'm not mistaken the external gateway only allows transfer operatorship to continue when the proof was signed by the current set: https://github.com/axelarnetwork/axelar-cgp-solidity/blob/main/contracts/AxelarGateway.sol#L502. We need to confirm the pending one before trying to confirm a new one or the new one will never be accepted by the external gateway. Is that right @cjcobb23 ?

Reply:
I'm just wondering if we change to revert, would there be a scenario where we could get the worker set rotation stuck if one set couldn't be confirmed or something

Reply:
I think we should revert, and add a TODO item to come up with a way to abort a worker set update if it gets stuck. I can't think of a reason it would get stuck, and we can always migrate the contract if there is some reason that it does. But maybe there is a cleaner way. Either way, I think for this PR, revert is fine, and circle back later.

We should allow resigning the same worker set update though. I could see the proof id getting lost, or the signing event being missed for whatever reason, and it would be useful to be able to retrigger the event. So this would mean only revert if `NEXT_WORKER_SET` is not empty, and the stored value is different than the value we are trying to store. If this ends up being complicated, I'm ok leaving this for another PR too.

Reply:
changed to revert